### PR TITLE
hugo/feature/Improve MotionKit

### PR DESCRIPTION
**Pull request by HPezz**
Improve smart mobile reinforcer 1080 rotation and add timeout to Motion Kit to prevent infinite rotation

- [x] Validated on robot

### CODE REVIEW

File: libs/MotionKit/include/MotionKit.h [hunk 1]
```diff
@@ -16,8 +16,12 @@ class MotionKit
...
   public:
 	MotionKit(interface::Motor &motor_left, interface::Motor &motor_right, IMUKit &imu_kit,
-			  interface::EventLoop &event_loop)
-		: _motor_left(motor_left), _motor_right(motor_right), _imukit(imu_kit), _event_loop(event_loop) {};
+			  interface::EventLoop &event_loop, interface::Timeout &timeout)
+		: _motor_left(motor_left),
+		  _motor_right(motor_right),
+		  _imukit(imu_kit),
+		  _event_loop(event_loop),
+		  _timeout(timeout) {};
```
Comment (ladislas) [inline libs/MotionKit/include/MotionKit.h:24]: having a `;` is an error

```suggestion
		  _timeout(timeout) {}
```

File: libs/MotionKit/include/MotionKit.h [hunk 2]
```diff
@@ -47,9 +52,9 @@ class MotionKit
 	bool _rotate_x_turns_requested = false;
 
 	const float kReferenceAngle		   = 180.F;
-	const float kMinimalViableRobotPwm = 0.25F;	  // ? Under this pwm value, torque is too low to spin the wheel
+	const float kMinimalViableRobotPwm = 0.15F;	  // ? Under this pwm value, torque is too low to spin the wheel
 	const float kPwmMaxValue		   = 1.F;
-	const float kPwmMarginLimit		   = 0.28F;
+	const float kPwmMarginLimit		   = 0.17F;
```
Comment (ladislas) [inline libs/MotionKit/include/MotionKit.h:57]: why those values? what changed?

File: libs/MotionKit/source/MotionKit.cpp [hunk 3]
```diff
@@ -106,7 +111,7 @@ void MotionKit::run()
 
 auto MotionKit::mapSpeed(float speed) const -> float
 {
-	return utils::math::map(speed, 0.F, 1.F, kMinimalViableRobotPwm, kPwmMaxValue);
+	return utils::math::map(speed, 0.F, 1.8F, kMinimalViableRobotPwm, kPwmMaxValue);
```
Comment (ladislas) [inline libs/MotionKit/source/MotionKit.cpp:114]: why this change? you'll get PWM values > 1.0, no?

**Review by ladislas — CHANGES_REQUESTED** [review]
looks good, but I have a few questions on the values changed.

has this been tested on a closed robot -- I mean **really** closed

File: libs/MotionKit/source/MotionKit.cpp (hunk 3 above)
Comment (YannLocatelli) [reply libs/MotionKit/source/MotionKit.cpp:114]: Oh, you should set it to the maximum value of PID if it is instead of magic number `1.8F`

File: libs/MotionKit/include/MotionKit.h (hunk 2 above)
Comment (YannLocatelli) [reply libs/MotionKit/include/MotionKit.h:57]: Yes, why this is more performant than last time? Because the robot is closed?

File: libs/MotionKit/source/MotionKit.cpp [hunk 4]
```diff
@@ -43,6 +43,11 @@ void MotionKit::rotate(uint8_t number_of_rotations, Rotation direction,
 	_motor_left.spin(direction, kPwmMaxValue);
 	_motor_right.spin(direction, kPwmMaxValue);
 
+	auto on_timeout = [this] { stop(); };
+
+	_timeout.onTimeout(on_timeout);
+	_timeout.start(10s);
```
Comment (YannLocatelli) [inline libs/MotionKit/source/MotionKit.cpp:49]: `_timeout.stop()` in `MotionKit::stop` too. Otherwise it motion will be too short for two Reinforcer in a row.

File: libs/MotionKit/tests/MotionKit_test.cpp [hunk 5]
```diff
@@ -36,12 +37,14 @@ class MotionKitTest : public ::testing::Test
 	mock::Accelerometer accel {};
 	mock::Gyroscope gyro {};
 
+	mock::Timeout mock_timeout {};
```
Comment (YannLocatelli) [inline libs/MotionKit/tests/MotionKit_test.cpp:40]: Add a test using the timeout callback to ensure you stop the motion at end of the timeout

**Review by YannLocatelli — CHANGES_REQUESTED** [review]

File: libs/MotionKit/include/MotionKit.h [hunk 6]
```diff
@@ -51,10 +53,15 @@ class MotionKit
 	bool _stabilisation_requested  = false;
 	bool _rotate_x_turns_requested = false;
 
-	const float kReferenceAngle		   = 180.F;
-	const float kMinimalViableRobotPwm = 0.15F;	  // ? Under this pwm value, torque is too low to spin the wheel
+	const float kReferenceAngle = 180.F;
+	const float kPIDMaxValue	= 1.8F;
+
+	// ? Under this pwm value, torque is too low to spin the wheel without inertia.
+	// ? This value needs to be the lowest possible to prevent inertia drift after reaching target
```
Comment (ladislas) [inline libs/MotionKit/include/MotionKit.h:60]:
```suggestion
	// ? When the motor is stopped, PWM values under kMinimalViableRobotPwm are too low to generate enough torque for the motor to start spinning
	// ? At the same time, kMinimalViableRobotPwm needs to be the lowest possible to avoid overshooting when the target is reached
```

**Review by YannLocatelli — APPROVED** [review]

**Review by ladislas — APPROVED** [review]
